File: src-tauri/src/commands/twitch/moderation.rs
use crate::dto::twitch::moderation::{Ban, BannedUser};
use crate::dto::twitch::user::UserRef;
use crate::services;
use crate::dto::pagination::PaginatedResponse;
use crate::services::twitch::get_token;
use serde::Deserialize;

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteChatMessagesParams {
    pub broadcaster_id: String,
    pub message_id: Option<String>,
}

#[tauri::command]
pub async fn delete_chat_messages(
    app: tauri::AppHandle,
    params: DeleteChatMessagesParams,
) -> Result<(), String> {
    let token = get_token(&app).await?;
    services::twitch::moderation::delete_chat_messages(
        &token,
        params.broadcaster_id,
        params.message_id,
    )
    .await
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BanUserParams {
    pub broadcaster_id: String,
    pub user_id: String,
    #[serde(default)]
    pub duration: Option<u32>,
    #[serde(default)]
    pub reason: Option<String>,
}

#[tauri::command]
pub async fn ban_user(app: tauri::AppHandle, params: BanUserParams) -> Result<Ban, String> {
    let token = get_token(&app).await?;
    services::twitch::moderation::ban_user(
        &token,
        params.broadcaster_id,
        params.user_id,
        params.duration,
        params.reason,
    )
    .await
}

#[derive(Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct GetBannedUsersParams {
    pub broadcaster_id: String,
    pub user_id: Option<String>,
    pub first: Option<usize>,
    pub after: Option<String>,
}

#[tauri::command]
pub async fn get_banned_users(
    app: tauri::AppHandle,
    params: GetBannedUsersParams,
) -> Result<PaginatedResponse<BannedUser>, String> {
    let token = get_token(&app).await?;
    services::twitch::moderation::get_banned_users(
        &token,
        params.broadcaster_id,
        params.user_id,
        params.first,
        params.after,
    )
    .await
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UnbanUserParams {
    pub broadcaster_id: String,
    pub user_id: String,
}

#[tauri::command]
pub async fn unban_user(app: tauri::AppHandle, params: UnbanUserParams) -> Result<(), String> {
    let token = get_token(&app).await?;
    services::twitch::moderation::unban_user(&token, params.broadcaster_id, params.user_id).await
}

#[derive(Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct GetModeratorsParams {
    pub broadcaster_id: String,
    pub first: Option<usize>,
    pub after: Option<String>,
}

#[tauri::command]
pub async fn get_moderators(
    app: tauri::AppHandle,
    params: GetModeratorsParams,
) -> Result<PaginatedResponse<UserRef>, String> {
    let token = get_token(&app).await?;
    services::twitch::moderation::get_moderators(
        &token,
        params.broadcaster_id,
        params.first,
        params.after,
    )
    .await
}

#[derive(Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct GetModeratedChannelsParams {
    pub first: Option<usize>,
    pub after: Option<String>,
}

#[tauri::command]
pub async fn get_moderated_channels(
    app: tauri::AppHandle,
    params: GetModeratedChannelsParams,
) -> Result<PaginatedResponse<UserRef>, String> {
    let token = get_token(&app).await?;
    services::twitch::moderation::get_moderated_channels(&token, params.first, params.after).await
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WarnUserParams {
    pub broadcaster_id: String,
    pub user_id: String,
    pub reason: String,
}

#[tauri::command]
pub async fn warn_user(app: tauri::AppHandle, params: WarnUserParams) -> Result<(), String> {
    let token = get_token(&app).await?;
    services::twitch::moderation::warn_user(
        &token,
        params.broadcaster_id,
        params.user_id,
        params.reason,
    )
    .await
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ManageHeldAutomodMessageParams {
    pub msg_id: String,
}

#[tauri::command]
pub async fn approve_held_automod_message(
    app: tauri::AppHandle,
    params: ManageHeldAutomodMessageParams,
) -> Result<(), String> {
    let token = get_token(&app).await?;
    services::twitch::moderation::manage_held_automod_message(&token, params.msg_id, true).await
}

#[tauri::command]
pub async fn deny_held_automod_message(
    app: tauri::AppHandle,
    params: ManageHeldAutomodMessageParams,
) -> Result<(), String> {
    let token = get_token(&app).await?;
    services::twitch::moderation::manage_held_automod_message(&token, params.msg_id, false).await
}

File: src/lib/api/twitch/moderation.ts
import type { Ban, BannedUser } from "../../types/twitch/moderation.ts";
import type { UserRef } from "../../types/twitch/user.ts";
import { invokeCommand, type InvokeOptions, type PaginatedResponse } from "../utils.ts";

export type { Ban, BannedUser } from "../../types/twitch/moderation.ts";

export type DeleteChatMessagesParams = {
  broadcasterId: string;
  messageId?: string | null;
};

export function deleteChatMessages(
  params: DeleteChatMessagesParams,
  options?: InvokeOptions,
): Promise<void> {
  const successMessage = params.messageId ? "Message deleted" : "Chat cleared";
  return invokeCommand("delete_chat_messages", params, { successMessage, ...options });
}

export type BanUserParams = {
  broadcasterId: string;
  userId: string;
  duration?: number | null;
  reason?: string | null;
};

export function banUser(
  params: BanUserParams,
  options?: InvokeOptions,
): Promise<Ban> {
  const successMessage = params.duration ? "User timed out" : "User banned";
  return invokeCommand("ban_user", params, { successMessage, ...options });
}

export type UnbanUserParams = {
  broadcasterId: string;
  userId: string;
};

export function unbanUser(
  params: UnbanUserParams,
  options?: InvokeOptions,
): Promise<void> {
  return invokeCommand("unban_user", params, { successMessage: "User unbanned", ...options });
}

export type GetBannedUsersParams = {
  broadcasterId: string;
  userId?: string;
  first?: number;
  after?: string;
};

export function getBannedUsers(
  params: GetBannedUsersParams,
  options?: InvokeOptions,
): Promise<PaginatedResponse<BannedUser>> {
  return invokeCommand("get_banned_users", params, options);
}

export type GetModeratorsParams = {
  broadcasterId: string;
  first?: number;
  after?: string;
};

export function getModerators(
  params: GetModeratorsParams,
  options?: InvokeOptions,
): Promise<PaginatedResponse<UserRef>> {
  return invokeCommand("get_moderators", params, options);
}

export type GetModeratedChannelsParams = {
  first?: number;
  after?: string;
};

export function getModeratedChannels(
  params: GetModeratedChannelsParams = {},
  options?: InvokeOptions,
): Promise<PaginatedResponse<UserRef>> {
  return invokeCommand("get_moderated_channels", params, options);
}

export type WarnUserParams = {
  broadcasterId: string;
  userId: string;
  reason: string;
};

export function warnUser(
  params: WarnUserParams,
  options?: InvokeOptions,
): Promise<void> {
  return invokeCommand("warn_user", params, { successMessage: "User warned", ...options });
}

export type ManageHeldAutomodMessageParams = {
  msgId: string;
};

export function approveHeldAutomodMessage(
  params: ManageHeldAutomodMessageParams,
  options?: InvokeOptions,
): Promise<void> {
  return invokeCommand("approve_held_automod_message", params, {
    successMessage: "Message approved",
    ...options,
  });
}

export function denyHeldAutomodMessage(
  params: ManageHeldAutomodMessageParams,
  options?: InvokeOptions,
): Promise<void> {
  return invokeCommand("deny_held_automod_message", params, {
    successMessage: "Message denied",
    ...options,
  });
}
